posts/2021/10/hpy-0.0.3.md

```diff
@@ -0,0 +1,188 @@
+
+```
+$ pypy -m pip show hpy
+Name: hpy
+Version: 0.0.3
+...
+
+$ graalpython -m pip show hpy
+Name: hpy
+Version: 0.0.3
+...
+
+$ pypy -c 'import hpy.universal; print(hpy.universal.get_version()[0])'
+0.0.3
+
+$ graalpython -c 'import hpy.universal; print(hpy.universal.get_version()[0])'
+0.0.3
+```
+
+API
+===
+
+HPy is still under heavy development and does not yet provide as many API
+functions as Python's C API but more and more pieces are added.
+In this release, we've added following functions:
+
+  - `HPy_Is`
+  - `HPyBytes_FromStringAndSize`
+  - `HPyErr_NewException`
+  - `HPyErr_NewExceptionWithDoc`
+  - `HPyUnicode_AsUTF8AndSize`
+  - `HPyUnicode_DecodeFSDefault`
+  - `HPyImport_ImportModule`
+
+All of the above functions except of `HPy_Is` have equivalents in the C API. You
+just need to remove the leading `H` from the name.
+
+Function `HPy_Is` was added to be able to do object identity comparisons. In the
+C API, you would just do a pointer comparison like this:
+
+```
+PyObject *object_a = /* ... */;
+PyObject *object_b = /* ... */;
+return object_a == object_b;
+```
+
+Since HPy was deliberately designed to prevent such comparisons, you need to use
+the newly introduced function:
+
+```
+HPy object_a = /* ... */;
+HPy object_b = /* ... */;
+return HPy_Is(ctx, object_a, object_b);
+```
+
+We've also extended functions `HPyArg_Parse` and `HPyArg_ParseKeywords` and
+added support for format specifier `s`. The new specifier converts a Python
+unicode object into a C string. For example:
+
+```
+const char *result;
+if (!HPyArg_Parse(ctx, NULL, args, nargs, "s", &result)) {
+    return HPy_NULL;
+}
+printf("result = %s\n", result);
+```
+
+For an overview of the current API, please refer to the public API declaration
+in [`public_api.h`](https://github.com/hpyproject/hpy/blob/release/0.0.3/hpy/tools/autogen/public_api.h#L107-L298),
+which is used to autogenerate parts of the HPy code and is a reliable list of
+all the supported functions.
+
+!!! Warning
+
+    The HPy API is still considered in alpha status and it's subject to change
+    between versions.
+
+Debug Mode
+===
+
+One of the great features of HPy is the _debug mode_. It was already introduced
+a while ago (and therefore already include in release 0.0.2) but it got two new
+features that can help C extension developers a lot.
+HPy's debug mode is now able to track closed handles and users now have the
+possibility to register a callback that is invoked when an invalid handle is
+accessed.
+
+For example, if you would access a close handle, the debug mode reports that and
+by default such errors are fatal:
+
+```
+HPyDef_METH(f, "f", f_impl, HPyFunc_O)
+static HPy f_impl(HPyContext *ctx, HPy self, HPy arg)
+{
+    HPy h = HPy_Dup(ctx, arg);
+    HPy_Close(ctx, h);
+    /* The following close will operate on an invalid handle and thus cause a
+    fatal error */
+    HPy_Close(ctx, h);
+    return HPy_Dup(ctx, ctx->h_None);
+}
+```
+
```

Comment:
is it actually true? I think that at the moment, a double-close doesn't crash, see https://github.com/hpyproject/hpy/issues/258#issuecomment-958785897

The crash happens if you try to actually use a closed handle.
Note that I am more and more convinced that allowing multiple `HPy_Close` was a mistake and that we should change the current behavior.

Reply:
I've removed this example since it does no longer make sense.